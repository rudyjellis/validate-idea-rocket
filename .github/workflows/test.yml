name: Test

on:
  push:
    branches: [main, develop]
  pull_request:
    branches: [main, develop]

jobs:
  test:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x, 20.x]

    steps:
      - uses: actions/checkout@v4

      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ matrix.node-version }}
          cache: 'npm'

      - name: Install dependencies
        run: npm ci

      - name: Run linter
        run: npm run lint

      - name: Run tests
        run: npm test

      - name: Generate coverage report
        run: npm run test:coverage

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v3
        with:
          files: ./coverage/coverage-final.json
          flags: unittests
          name: codecov-umbrella
