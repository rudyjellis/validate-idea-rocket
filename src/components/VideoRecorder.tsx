import React, { useRef, useState, useEffect } from "react";
import { Button } from "@/components/ui/button";
import { useToast } from "@/components/ui/use-toast";
import { Video, StopCircle, Camera } from "lucide-react";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";

const VideoRecorder = () => {
  const videoRef = useRef<HTMLVideoElement>(null);
  const mediaRecorderRef = useRef<MediaRecorder | null>(null);
  const streamRef = useRef<MediaStream | null>(null);
  const [isRecording, setIsRecording] = useState(false);
  const [recordedChunks, setRecordedChunks] = useState<Blob[]>([]);
  const [timeLeft, setTimeLeft] = useState(30);
  const [cameras, setCameras] = useState<MediaDeviceInfo[]>([]);
  const [selectedCamera, setSelectedCamera] = useState<string>("");
  const { toast } = useToast();
  const MAX_RECORDING_TIME = 30000; // 30 seconds in milliseconds

  useEffect(() => {
    // Get available cameras on component mount
    const getCameras = async () => {
      try {
        const devices = await navigator.mediaDevices.enumerateDevices();
        const videoDevices = devices.filter(device => device.kind === 'videoinput');
        setCameras(videoDevices);
        if (videoDevices.length > 0) {
          setSelectedCamera(videoDevices[0].deviceId);
        }
      } catch (error) {
        console.error("Error getting cameras:", error);
        toast({
          variant: "destructive",
          title: "Error",
          description: "Could not get available cameras",
        });
      }
    };

    getCameras();
  }, []);

  useEffect(() => {
    return () => {
      // Cleanup function to stop all media tracks
      if (streamRef.current) {
        streamRef.current.getTracks().forEach(track => track.stop());
      }
    };
  }, []);

  useEffect(() => {
    let intervalId: NodeJS.Timeout;
    
    if (isRecording) {
      setTimeLeft(30);
      intervalId = setInterval(() => {
        setTimeLeft((prevTime) => {
          if (prevTime <= 1) {
            clearInterval(intervalId);
            return 0;
          }
          return prevTime - 1;
        });
      }, 1000);
    }

    return () => {
      if (intervalId) {
        clearInterval(intervalId);
      }
    };
  }, [isRecording]);

  const startRecording = async () => {
    try {
      // Stop any existing stream
      if (streamRef.current) {
        streamRef.current.getTracks().forEach(track => track.stop());
      }

      const stream = await navigator.mediaDevices.getUserMedia({ 
        video: { deviceId: selectedCamera },
        audio: true 
      });
      
      streamRef.current = stream;
      if (videoRef.current) {
        videoRef.current.srcObject = stream;
      }

      const mediaRecorder = new MediaRecorder(stream);
      mediaRecorderRef.current = mediaRecorder;
      
      const chunks: Blob[] = [];
      mediaRecorder.ondataavailable = (e) => {
        if (e.data.size > 0) {
          chunks.push(e.data);
        }
      };

      mediaRecorder.onstop = () => {
        setRecordedChunks(chunks);
        if (streamRef.current) {
          streamRef.current.getTracks().forEach(track => track.stop());
        }
        setIsRecording(false);
      };

      mediaRecorder.start();
      setIsRecording(true);
      
      // Automatically stop recording after MAX_RECORDING_TIME
      setTimeout(() => {
        if (mediaRecorder.state === "recording") {
          stopRecording();
          toast({
            title: "Recording completed",
            description: "Maximum recording time (30 seconds) reached",
          });
        }
      }, MAX_RECORDING_TIME);

    } catch (error) {
      console.error("Error accessing media devices:", error);
      toast({
        variant: "destructive",
        title: "Error",
        description: "Could not access camera or microphone",
      });
    }
  };

  const stopRecording = () => {
    if (mediaRecorderRef.current?.state === "recording") {
      mediaRecorderRef.current.stop();
    }
  };

  const downloadVideo = () => {
    if (recordedChunks.length === 0) return;
    
    const blob = new Blob(recordedChunks, { type: "video/webm" });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    document.body.appendChild(a);
    a.style.display = "none";
    a.href = url;
    a.download = "recorded-video.webm";
    a.click();
    URL.revokeObjectURL(url);
    document.body.removeChild(a);
  };

  const handleCameraChange = async (deviceId: string) => {
    setSelectedCamera(deviceId);
    if (isRecording) {
      stopRecording();
    }
    
    try {
      if (streamRef.current) {
        streamRef.current.getTracks().forEach(track => track.stop());
      }
      
      const newStream = await navigator.mediaDevices.getUserMedia({
        video: { deviceId },
        audio: true
      });
      
      streamRef.current = newStream;
      if (videoRef.current) {
        videoRef.current.srcObject = newStream;
      }
    } catch (error) {
      console.error("Error switching camera:", error);
      toast({
        variant: "destructive",
        title: "Error",
        description: "Could not switch camera",
      });
    }
  };

  return (
    <div className="flex flex-col items-center gap-4 p-4">
      <div className="w-full max-w-md">
        {cameras.length > 1 && (
          <div className="mb-4">
            <Select
              value={selectedCamera}
              onValueChange={handleCameraChange}
              disabled={isRecording}
            >
              <SelectTrigger className="w-full">
                <SelectValue placeholder="Select camera" />
              </SelectTrigger>
              <SelectContent>
                {cameras.map((camera) => (
                  <SelectItem key={camera.deviceId} value={camera.deviceId}>
                    <div className="flex items-center gap-2">
                      <Camera className="h-4 w-4" />
                      {camera.label || `Camera ${cameras.indexOf(camera) + 1}`}
                    </div>
                  </SelectItem>
                ))}
              </SelectContent>
            </Select>
          </div>
        )}
        
        <div className="relative">
          <video
            ref={videoRef}
            autoPlay
            playsInline
            muted
            className="w-full rounded-lg border border-gray-200 bg-black"
          />
          {isRecording && (
            <div className="absolute top-4 right-4 bg-black/75 text-white px-3 py-1 rounded-full">
              {timeLeft}s
            </div>
          )}
        </div>
      </div>
      
      <div className="flex gap-2">
        {!isRecording ? (
          <Button
            onClick={startRecording}
            className="gap-2"
            disabled={recordedChunks.length > 0}
          >
            <Video className="h-4 w-4" />
            Start Recording
          </Button>
        ) : (
          <Button
            onClick={stopRecording}
            variant="destructive"
            className="gap-2"
          >
            <StopCircle className="h-4 w-4" />
            Stop Recording
          </Button>
        )}
        
        {recordedChunks.length > 0 && (
          <Button onClick={downloadVideo} variant="outline">
            Download Recording
          </Button>
        )}
      </div>
    </div>
  );
};

export default VideoRecorder;