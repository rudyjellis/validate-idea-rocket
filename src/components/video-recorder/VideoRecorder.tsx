import { useState, useEffect } from "react";
import type { VideoRecorderProps } from "./types";
import CameraSelector from "./CameraSelector";
import VideoPreview from "./VideoPreview";
import RecordingControls from "./RecordingControls";
import { useVideoRecording } from "./hooks/useVideoRecording";
import { useCameraDevices } from "./hooks/useCameraDevices";
import { useIsMobile } from "@/hooks/use-mobile";

const VideoRecorder = ({ maxDuration = 30, onRecordingComplete }: VideoRecorderProps) => {
  const [isPlayingBack, setIsPlayingBack] = useState(false);
  const isMobile = useIsMobile();
  
  const {
    videoRef,
    recordingState,
    recordedChunks,
    timeLeft,
    startRecording,
    stopRecording,
    pauseRecording,
    resumeRecording,
    initializeStream,
    downloadVideo,
    playRecording,
  } = useVideoRecording();

  const { cameras, selectedCamera, setSelectedCamera } = useCameraDevices();

  useEffect(() => {
    if (selectedCamera) {
      console.log("Initializing stream for selected camera:", selectedCamera);
      initializeStream(selectedCamera);
    }
  }, [selectedCamera, initializeStream]);

  const handleCameraChange = (deviceId: string) => {
    console.log("Camera changed to:", deviceId);
    setSelectedCamera(deviceId);
    if (recordingState !== "idle") {
      stopRecording();
    }
  };

  const handleStartRecording = () => {
    console.log("Starting recording with camera:", selectedCamera);
    setIsPlayingBack(false);
    startRecording(selectedCamera);
  };

  const handlePlayback = () => {
    console.log("Starting playback");
    setIsPlayingBack(true);
    playRecording();
    if (videoRef.current) {
      videoRef.current.onended = () => {
        console.log("Playback ended");
        setIsPlayingBack(false);
      };
    }
  };

  const handleTapToRecord = () => {
    console.log("Tap to record triggered");
    if (recordingState === "idle") {
      handleStartRecording();
    } else if (recordingState === "recording") {
      stopRecording();
    }
  };

  return (
    <div className={`flex flex-col ${isMobile ? 'h-full' : ''}`}>
      <div className={`w-full ${isMobile ? 'flex-1 relative' : 'max-w-md mx-auto'}`}>
        {(!isMobile || recordingState === "idle") && (
          <CameraSelector
            cameras={cameras}
            selectedCamera={selectedCamera}
            onCameraChange={handleCameraChange}
            disabled={recordingState !== "idle"}
          />
        )}
        
        <div className={`${isMobile ? 'absolute inset-0' : 'mt-4'}`}>
          <VideoPreview
            ref={videoRef}
            isRecording={recordingState === "recording"}
            timeLeft={timeLeft}
            recordingState={recordingState}
            isPlayingBack={isPlayingBack}
            onTapToRecord={handleTapToRecord}
          />
        </div>
      </div>

      <div className={`${isMobile ? 'fixed bottom-0 left-0 right-0 pb-6 bg-gradient-to-t from-black/80 to-transparent pt-20' : ''}`}>
        <RecordingControls
          recordingState={recordingState}
          onStartRecording={handleStartRecording}
          onStopRecording={stopRecording}
          onPauseRecording={pauseRecording}
          onResumeRecording={resumeRecording}
          onDownload={downloadVideo}
          onPlayback={handlePlayback}
          hasRecording={recordedChunks.length > 0}
        />
      </div>
    </div>
  );
};

export default VideoRecorder;