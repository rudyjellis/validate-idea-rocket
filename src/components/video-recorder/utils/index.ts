import { type VideoFormat } from "../types";

export const formatTime = (seconds: number): string => {
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = Math.floor(seconds % 60);
  return `${minutes}:${remainingSeconds.toString().padStart(2, "0")}`;
};

export const initializeMediaStream = async (selectedCamera: string) => {
  try {
    console.log("Initializing media stream for camera:", selectedCamera);
    const stream = await navigator.mediaDevices.getUserMedia({
      video: { deviceId: selectedCamera },
      audio: true,
    });
    console.log("Media stream initialized successfully");
    return stream;
  } catch (error) {
    console.error("Error accessing media devices:", error);
    throw error;
  }
};

export const stopMediaStream = (stream: MediaStream | null) => {
  if (stream) {
    stream.getTracks().forEach((track) => track.stop());
    console.log("Media stream stopped");
  }
};

export const createDownloadLink = (
  recordedChunks: Blob[],
  format: VideoFormat,
  counter: number
): { url: string; fileName: string } => {
  const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
  const fileName = `recording-${timestamp}-${counter}.${format}`;
  const mimeType = format === "webm" ? "video/webm" : "video/mp4";
  const blob = new Blob(recordedChunks, { type: mimeType });
  const url = URL.createObjectURL(blob);
  
  return { url, fileName };
};