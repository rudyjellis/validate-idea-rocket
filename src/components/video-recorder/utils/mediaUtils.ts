export const initializeMediaStream = async (selectedCamera: string) => {
  try {
    const stream = await navigator.mediaDevices.getUserMedia({
      video: { deviceId: selectedCamera },
      audio: true,
    });
    return stream;
  } catch (error) {
    console.error("Error accessing media devices:", error);
    throw error;
  }
};

export const stopMediaStream = (stream: MediaStream | null) => {
  if (stream) {
    stream.getTracks().forEach((track) => track.stop());
  }
};