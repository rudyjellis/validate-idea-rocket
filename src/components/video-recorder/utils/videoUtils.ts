import { VideoFormat } from "../types/video";
import { toast } from "@/hooks/use-toast";

export const createVideoUrl = (chunks: Blob[]): string => {
  const blob = new Blob(chunks, { type: 'video/webm' });
  return URL.createObjectURL(blob);
};

export const downloadVideoFile = (chunks: Blob[], format: VideoFormat) => {
  if (chunks.length === 0) return;

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const fileName = `recording-${timestamp}`;
  const mimeType = format === 'webm' ? 'video/webm' : 'video/mp4';
  const blob = new Blob(chunks, { type: mimeType });
  const url = URL.createObjectURL(blob);
  
  const a = document.createElement("a");
  document.body.appendChild(a);
  a.style.display = "none";
  a.href = url;
  a.download = `${fileName}.${format}`;
  a.click();
  
  URL.revokeObjectURL(url);
  document.body.removeChild(a);

  toast({
    title: "Download started",
    description: `Your video will be downloaded in ${format.toUpperCase()} format`,
  });
};