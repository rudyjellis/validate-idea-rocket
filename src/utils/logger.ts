/**
 * Conditional logger utility that only logs in development mode
 * Helps improve performance in production by removing console calls
 */

type LogLevel = 'log' | 'warn' | 'error' | 'info' | 'debug';

const isDevelopment = process.env.NODE_ENV === 'development';

export const logger = {
  log: (...args: any[]) => {
    if (isDevelopment) {
      console.log(...args);
    }
  },
  
  warn: (...args: any[]) => {
    if (isDevelopment) {
      console.warn(...args);
    }
  },
  
  error: (...args: any[]) => {
    if (isDevelopment) {
      console.error(...args);
    }
  },
  
  info: (...args: any[]) => {
    if (isDevelopment) {
      console.info(...args);
    }
  },
  
  debug: (...args: any[]) => {
    if (isDevelopment) {
      console.debug(...args);
    }
  }
};

// For video recorder specific logging with context
export const createVideoRecorderLogger = (context: string) => ({
  log: (message: string, ...args: any[]) => {
    if (isDevelopment) {
      console.log(`[${context}] ${message}`, ...args);
    }
  },
  
  warn: (message: string, ...args: any[]) => {
    if (isDevelopment) {
      console.warn(`[${context}] ${message}`, ...args);
    }
  },
  
  error: (message: string, ...args: any[]) => {
    if (isDevelopment) {
      console.error(`[${context}] ${message}`, ...args);
    }
  },
  
  info: (message: string, ...args: any[]) => {
    if (isDevelopment) {
      console.info(`[${context}] ${message}`, ...args);
    }
  },
  
  debug: (message: string, ...args: any[]) => {
    if (isDevelopment) {
      console.debug(`[${context}] ${message}`, ...args);
    }
  }
});

