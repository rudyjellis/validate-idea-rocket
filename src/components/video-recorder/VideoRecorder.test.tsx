import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render } from '@/test/utils';
import VideoRecorder from './VideoRecorder';

// Mock the mobile/desktop components
vi.mock('./components/MobileVideoRecorder', () => ({
  default: ({ maxDuration }: { maxDuration: number }) => (
    <div data-testid="mobile-video-recorder" data-max-duration={maxDuration}>
      Mobile Video Recorder
    </div>
  ),
}));

vi.mock('./components/DesktopVideoRecorder', () => ({
  default: ({ maxDuration }: { maxDuration: number }) => (
    <div data-testid="desktop-video-recorder" data-max-duration={maxDuration}>
      Desktop Video Recorder
    </div>
  ),
}));

// Mock useIsMobile hook
const mockUseIsMobile = vi.fn();
vi.mock('@/hooks/use-mobile', () => ({
  useIsMobile: () => mockUseIsMobile(),
}));

describe('VideoRecorder Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Desktop Mode', () => {
    beforeEach(() => {
      mockUseIsMobile.mockReturnValue(false);
    });

    it('renders desktop video recorder when not mobile', () => {
      const { getByTestId } = render(<VideoRecorder />);
      expect(getByTestId('desktop-video-recorder')).toBeInTheDocument();
    });

    it('passes maxDuration prop to desktop recorder', () => {
      const { getByTestId } = render(<VideoRecorder maxDuration={60} />);
      const recorder = getByTestId('desktop-video-recorder');
      expect(recorder).toHaveAttribute('data-max-duration', '60');
    });

    it('uses default maxDuration of 30 seconds', () => {
      const { getByTestId } = render(<VideoRecorder />);
      const recorder = getByTestId('desktop-video-recorder');
      expect(recorder).toHaveAttribute('data-max-duration', '30');
    });

    it('matches desktop snapshot', () => {
      const { container } = render(<VideoRecorder />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Mobile Mode', () => {
    beforeEach(() => {
      mockUseIsMobile.mockReturnValue(true);
    });

    it('renders mobile video recorder when on mobile', () => {
      const { getByTestId } = render(<VideoRecorder />);
      expect(getByTestId('mobile-video-recorder')).toBeInTheDocument();
    });

    it('passes maxDuration prop to mobile recorder', () => {
      const { getByTestId } = render(<VideoRecorder maxDuration={45} />);
      const recorder = getByTestId('mobile-video-recorder');
      expect(recorder).toHaveAttribute('data-max-duration', '45');
    });

    it('applies mobile-specific styling', () => {
      const { container } = render(<VideoRecorder />);
      const wrapper = container.querySelector('.h-\\[100dvh\\]');
      expect(wrapper).toBeInTheDocument();
    });

    it('matches mobile snapshot', () => {
      const { container } = render(<VideoRecorder />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Error Boundary', () => {
    it('wraps component in error boundary', () => {
      const { container } = render(<VideoRecorder />);
      // Error boundary should be present in the component tree
      expect(container.firstChild).toBeTruthy();
    });
  });

  describe('Component Structure', () => {
    beforeEach(() => {
      mockUseIsMobile.mockReturnValue(false);
    });

    it('maintains consistent structure', () => {
      const { container } = render(<VideoRecorder />);
      
      // Check for main wrapper
      const mainWrapper = container.querySelector('.flex.flex-col');
      expect(mainWrapper).toBeInTheDocument();
      
      // Check for inner container
      const innerContainer = container.querySelector('.w-full');
      expect(innerContainer).toBeInTheDocument();
    });

    it('applies transform-gpu for performance', () => {
      const { container } = render(<VideoRecorder />);
      const wrapper = container.querySelector('.transform-gpu');
      expect(wrapper).toBeInTheDocument();
    });

    it('applies transition classes', () => {
      const { container } = render(<VideoRecorder />);
      const innerContainer = container.querySelector('.transition-all');
      expect(innerContainer).toBeInTheDocument();
    });
  });
});
