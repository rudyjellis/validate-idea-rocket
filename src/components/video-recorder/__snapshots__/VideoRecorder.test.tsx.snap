// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VideoRecorder Component > Desktop Mode > matches desktop snapshot 1`] = `
<div
  class="
          flex 
          flex-col 
           
          transform-gpu
        "
>
  <div
    class="
            w-full 
            max-w-4xl mx-auto
            transition-all 
            duration-300 
            ease-in-out
          "
  >
    <div
      data-max-duration="30"
      data-testid="desktop-video-recorder"
    >
      Desktop Video Recorder
    </div>
  </div>
</div>
`;

exports[`VideoRecorder Component > Mobile Mode > matches mobile snapshot 1`] = `
<div
  class="
          flex 
          flex-col 
          h-[100dvh] 
          transform-gpu
        "
>
  <div
    class="
            w-full 
            flex-1 relative
            transition-all 
            duration-300 
            ease-in-out
          "
  >
    <div
      data-max-duration="30"
      data-testid="mobile-video-recorder"
    >
      Mobile Video Recorder
    </div>
  </div>
</div>
`;
